classes/local/model/cms_types.php

```diff
@@ -131,6 +135,52 @@ public function get_custom_data(string $name) {
         return $cdata->$name ?? null;
     }
 
+    /**
+     * Validates idnumber parameter.
+     *
+     * @param string $value
+     * @return bool|\lang_string
+     */
+    protected function validate_idnumber(string $value) {
```

Comment:
Someone once told me (I think it was @dmitriim) that even Moodle HQ frowns on using the validation methods inside persistents because they can cause more problems than they solve. So just a thought there.

Reply:
They frown on using their own code? I don't buy that. Can you point to a quote? Can you tell me what problems they cause?

Reply:
I read that post by dmittriim. It looks like he was referring to the whole persistant class itself, not just bits and pieces of it. Do you want to submit an issue to remove the persistant class?

If you truly object to my use of validate_xxx, then what alternative do you suggest. the extra_validation() method on the form only covers form submission. validate_xxx also covers in-code saving, which is why I moved it.

Reply:
Not a public one, no. We just got bitten with a bug that was caused by validation in persistent class so that's why I'm a bit wary. Hopefully it doesn't bite us as well.